File: Python/Experimenter.py
from Interface import Problem
import _datetime
import time
import timeit
import ExperimentResults


class Experimenter:



    def Avg(self, numbers):
        total = 0
        for i in numbers:
            total+=i
        return total/len(numbers)

    def Median(self, numbers):
        sortedList = sorted(numbers, key=float)
        return sortedList[int(len(sortedList)/2)]


    def High(self, numbers):
        sortedList = sorted(numbers, key=float)
        return sortedList[int(len(sortedList)-1)]

    def Low(self, numbers):
        sortedList = sorted(numbers, key=float)
        return sortedList[0]


    def RunExperiments(self, problem):
        print("Running experiment on: "+problem.__class__.__name__)
        timed = timeit.repeat(problem.Calculate, number=1, repeat=100000)
        avg = self.Avg(timed)*1000000000
        median = self.Median(timed)* 1000000000
        high = self.High(timed)*1000000000
        low = self.Low(timed) * 1000000000
        return ExperimentResults.Results(problem.__class__.__name__, timed, avg, median, high, low)


    def AvgOfAvg(self, results):
        total = 0
        for i in results:
            total += i.average
        return total/len(results)

    def AvgOfMedian(self, results):
        total = 0
        for i in results:
            total += i.median
        return total/len(results)

    def AvgOfHigh(self, results):
        total = 0
        for i in results:
            total += i.high
        return total/len(results)

    def AvgOfLow(self, results):
        total = 0
        for i in results:
            total += i.low
        return total/len(results)

    def PrintIndi(self, results):
        for i in results:
            i.PrintSelf()


    def PrintAvg(self, results):
        print("Printing Averages on: "+results[0].problemname)
        print("Average-Average: %.0fns" % self.AvgOfAvg(results))
        print("Average-Median: %.0fns" % self.AvgOfMedian(results))
        print("Average-High: %.0fns" % self.AvgOfHigh(results))
        print("Average-Low: %.0fns" % self.AvgOfLow(results))

File: Ruby/experimenter.rb
load 'problem.rb'
load 'experiment_results.rb'

class Experimenter

  def RunExperiment(problem)
    puts "Running experiment on: "+problem.class.name
    meassurements = []

    10.times do
      start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      result = problem.Calculate()
      endtime = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      elapsed = (endtime - start)*1000000000
      meassurements.push(elapsed)
      if (!result)
        print("The algorithm failed")
      end

    end

    return ExperimentResults.new(problem.class.name, meassurements, avg(meassurements), med(meassurements), high(meassurements), low(meassurements))

  end


  def avg(numbers)
    total = 0
    numbers.each {|num| total+=num}
    return total/numbers.size()
  end


  def med(numbers)
    numbers.sort
    return numbers[numbers.size()/2]
  end

  def high(numbers)
    numbers.sort
    return numbers[0]
  end

  def low(numbers)
    numbers.sort
    return numbers[numbers.size-1]
  end


  def PrintIndi(results)
    results.each {|i|
      i.PrintSelf
    }
  end

  def PrintAvg(results)
    puts "Printing Averages on: "+results[0].to_s+"ns"
    puts "Average-Average: "+AvgOfAvg(results).to_s+"ns"
    puts "Average-Median: "+AvgOfMed(results).to_s+"ns"
    puts "Average-High: " +AvgOfHigh(results).to_s+"ns"
    puts "Average-Low: " +AvgOfLow(results).to_s+"ns"
  end


  def AvgOfAvg(results)
    total = 0
    results.each {|i|
      total += i.avg
    }
    return (total/results.size).round

  end

  def AvgOfMed(results)
    total = 0
    results.each {|i|
      total += i.med
    }
    return (total/results.size).round

  end

  def AvgOfHigh(results)
    total = 0
    results.each {|i|
      total += i.high
    }
    return (total/results.size).round

  end

  def AvgOfLow(results)
    total = 0
    results.each {|i|
      total += i.low
    }
    return (total/results.size).round

  end



end